csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+            Fit a LambdaMART algorithm to the provided X and y arrays where X contains the features and y being the relevance scores.
+
+            Parameters
+            ----------
+            X : numpy array
+                (n_instances, n_objects, n_features)
+                Feature vectors of the objects
+            Y : numpy array
+                (n_instances, n_objects)
+                Rankings of the given objects
+            **kwargs
+                Keyword arguments for the fit function
+
+        """
+
+        training_data = self._prepare_train_data(X, y)
+        predicted_scores = np.zeros(len(training_data))
+        query_indexes = self._group_queries(training_data, 1)
+        query_keys = query_indexes.keys()
+        true_scores = [training_data[query_indexes[query], 0] for query in query_keys]
+
+        sorted_query_pair = []
+        for query_scores in true_scores:
+            temp = sorted(query_scores, reverse=True)
+            pairs = []
+            for i in range(len(temp)):
+                for j in range(len(temp)):
+                    if temp[i] > temp[j]:
+                        pairs.append((i,j))
+            sorted_query_pair.append(pairs)
+
+        idcg = [self._i_dcg(scores) for scores in true_scores]
+
+        for k in range(self.number_of_trees):
+            print("Tree %d" % (k))
+            lambdas = np.zeros(len(predicted_scores))
+            w = np.zeros(len(predicted_scores))
+            pred_scores = [predicted_scores[query_indexes[query]] for query in query_keys]
+            
+            pool = Pool()
+            for lambda_val, w_val, query_key in pool.map(compute_lambda_weights, zip(true_scores, pred_scores, sorted_query_pair, idcg, query_keys), chunksize=1):
+                indexes = query_indexes[query_key]
+                lambdas[indexes] = lambda_val
+                w[indexes] = w_val
+            pool.close()
+            # filename = str("lmartDebug2_"+str(k)+"_lambdas.txt")
+            # np.savetxt(filename, lambdas)
+
+            #Hyperparameters to be considered are :"min_samples_split","max_depth","min_samples_leaf","max_leaf_nodes"
+            tree = DecisionTreeRegressor(criterion="mse",
+                                         splitter="best",
+                                         max_depth=self.max_depth,
+                                         min_samples_split=self.min_samples_split,
+                                         min_samples_leaf=self.min_samples_leaf,
+                                         min_weight_fraction_leaf=0.,
+                                         max_features=None,
+                                         random_state=9,
+                                         max_leaf_nodes=self.max_leaf_nodes,
+                                         min_impurity_decrease=0.,
+                                         min_impurity_split=None)
+            tree.fit(training_data[:,2:], lambdas)
+            self.trees.append(tree)
```

Comment:
`self.trees` is not cleared - thus every time you call `fit`, it will keep appending new trees here.
This is also a nice candidate for a test.